File: Strohm/ios/Strohm/Sources/Strohm/ViewModels/KeyedArrayViewModel.swift
import Foundation

open class KeyedArrayViewModel<EntryType: ConstructableFromDictionary>: ViewModelBase<[EntryType]> {
    public var sorter: ((EntryType, EntryType) -> Bool)?

    public var entries: [EntryType] {
        willSet {
            // https://stackoverflow.com/questions/57615920/published-property-wrapper-not-working-on-subclass-of-observableobject
            self.objectWillChange.send()
        }
    }

    public init(initialEntries: [EntryType], propName: PropName, propPath: PropPath) {
        self.entries = initialEntries
        super.init(propName: propName, propPath: propPath)
    }

    override func propsToData(props: Props) -> [EntryType]? {
        guard let rawData = props[self.propName] as? [PropName: Props] else {
            return nil
        }

        var data = rawData.values.compactMap(EntryType.init(from:))
        if let sorter = self.sorter {
            data = data.sorted(by: sorter)
        }
        print("Received entries: ", data)
        return data
    }

    override func store(data: [EntryType]) {
        self.entries = data
    }
}

File: Strohm/android/strohm/src/main/java/dev/strohmnative/strohm/viewmodels/KeyedArrayViewModel.kt
package dev.strohmnative.strohm.viewmodels

import androidx.lifecycle.LiveData
import androidx.lifecycle.Transformations
import dev.strohmnative.strohm.*

open class KeyedArrayViewModel<EntryType>(
    initialData: List<EntryType>,
    propName: PropName,
    propPath: PropPath,
    private val instanceFactory: ConstructableFromDictionary<EntryType>
) : ViewModelBase<List<EntryType>>(initialData, propName, propPath) {
    var sorter: Comparator<EntryType>? = null
    var entries: LiveData<List<EntryType>> = Transformations.map(data) {
        data -> data as List<EntryType>
    }

    override fun propsToData(props: Props): List<EntryType>? {
        val m = props[this.propName] as? Map<*, *> ?: return null
        val rawData = m.asMapOfType<PropName, Props>() ?: return null
        var data = rawData.values.mapNotNull(instanceFactory::createFromDict)
        return sorter?.let { data.sortedWith(it) } ?: data
    }
}
